### ISSUE TITLE
[Due for payment 2025-02-28] Replace `react-native` Animated API with `react-native-reanimated`

### ISSUE THREAD

**Issue by blazejkustra** [Expensify/App#52920]
Full proposal [here](https://swmansion.slack.com/archives/C05LX9D6E07/p1732198006656509).

Replace Animated API with Reanimated, which runs animations on the UI thread (instead of JS thread), ensuring great performance. Reanimated integrates deeply with react-native-gesture-handler (used throughout our app), offers advanced animations, and supports higher frame rates — exceeding 60fps.

https://github.com/user-attachments/assets/dc578b3d-080b-4d1e-91ba-ff10c171abd2

<details><summary>Issue Owner</summary>Current Issue Owner: @stephanieelliott</details>

**Comment by mountiny** [comment]
Happy to help push this one along

**Comment by blazejkustra** [comment]
Update: I created a [draft PR](https://github.com/Expensify/App/pull/52956) with just the Switch component. I will get back to it next week, and involve @sumo-slonik to speed up the process 😄

**Comment by dubielzyk-expensify** [comment]
Oh boy this is wonderful. Is there any particular part of the app I can help review? 

Also, sidenote, but I would secretly love to add these kinda animations on that page if performance now permits this kinda stuff:

https://github.com/user-attachments/assets/43e48c18-6b18-4ec6-bc7f-3238f125d616

**Comment by blazejkustra** [comment]
> Also, sidenote, but I would secretly love to add these kinda animations on that page if performance now permits this kinda stuff:

That looks neat, we could look into this but perhaps as a separate issue? @dubielzyk-expensify

**Comment by mountiny** [comment]
Yeah we can do that in a separate issue for sure

**Comment by dubielzyk-expensify** [comment]
Sounds good to me 😄 This is excellent and it just got me a bit too excited haha 😅 60fps FTW

**Comment by mountiny** [comment]
sky is the limit, 120 fps 🚀

**Comment by sumo-slonik** [comment]
I have already managed to migrate the tooltip as well, I am now working on the tabs component and checking what could be improved in it :)

**Comment by sumo-slonik** [comment]
I migrated AttachmentModal and GrowlNotification

**Comment by mountiny** [comment]
`GrowlNotification` is not used in the app, is it? we should probably remove it

**Comment by sumo-slonik** [comment]
> `GrowlNotification` is not used in the app, is it? we should probably remove it
it seems to me that we can't remove it because there are places where we continue to show it:
![image](https://github.com/user-attachments/assets/b64ea693-e9aa-4fe5-bdc6-5664037e4a36)

**Comment by sumo-slonik** [comment]
@mountiny 
we also found the use of `react-native-animatable` would we like to swap that too with this PR?
<img width="1338" alt="image" src="https://github.com/user-attachments/assets/dc50713a-55d2-4f36-8539-f66fdea32253">

**Comment by sumo-slonik** [comment]
Update: `TextInput` is now ready

**Comment by mountiny** [comment]
@sumo-slonik @blazejkustra just to confirm, can we do this in some batches in multiple PRs? So we dont have one large PR with X components. This will help us catch regressions easier

**Comment by mountiny** [comment]
I think we will eventually remove it, but that is out of scope here, thanks for checking

**Comment by mountiny** [comment]
Lets leave that out for different issue, feel free to write a proposal in open source and tag design team

**Comment by blazejkustra** [comment]
First [PR](https://github.com/Expensify/App/pull/53298) is up! @shubham1206agra

**Comment by sumo-slonik** [comment]
PR for the second part is ready:
https://github.com/Expensify/App/pull/53321

**Comment by sumo-slonik** [comment]
> @blazejkustra I see a bug in recent switch PR.
> 
> Steps:
> 
> 1. Launch hybrid or ND app.
> 2. Create a Collect workspace.
> 3. Go to More features.
> 4. Enable Rules.
> 5. Close the upgrade modal without finishing.
> 6. Observe that the switch is on and it is stuck there.
> 
> Sorry, I didn't see this earlier.
> 
> cc @sumo-slonik

I see that a similar problem is also for credit cards, but I think I know how to fix it, I'm already dealing with it

**Comment by sumo-slonik** [comment]
@shubham1206agra
Would the expected behavior be for the switch to be false if we leave the update tab?

**Comment by sumo-slonik** [comment]
> > @blazejkustra I see a bug in recent switch PR.
> > Steps:
> > 
> > 1. Launch hybrid or ND app.
> > 2. Create a Collect workspace.
> > 3. Go to More features.
> > 4. Enable Rules.
> > 5. Close the upgrade modal without finishing.
> > 6. Observe that the switch is on and it is stuck there.
> > 
> > Sorry, I didn't see this earlier.
> > cc @sumo-slonik
> 

PR with bugfix: https://github.com/Expensify/App/pull/53379

**Comment by blazejkustra** [comment]
Let's wait with payments, second [PR](https://github.com/Expensify/App/pull/53321) is still not merged

**Comment by sumo-slonik** [comment]
@stephanieelliott proposed to add an animation after changing the switch, can you create an issiue with this item? I know how to do it and I could take care on it now.

**Comment by mountiny** [comment]
@sumo-slonik Sorry, can you please elaborate on what is the bug? Ideally, can you please post it in the #expensify-bugs channel in the expected format and then QA can create such issue

Or if this is a change request, we should also first discuss it with the team

**Comment by mountiny** [comment]
ah

**Comment by mountiny** [comment]
Here is an issue for that https://github.com/Expensify/App/issues/53759

@blazejkustra @sumo-slonik is this everything already migrated to reanimated? it feels like there should be more

**Comment by sumo-slonik** [comment]
@mountiny @blazejkustra 

Regarding the bug we discussed on Slack:
https://swmansion.slack.com/archives/C049HHMV9SM/p1733893876395249

I have identified the cause of the bug, which is located in the **react-native-draggable-flatlist** library. After investigating, I discovered an open issue related to this problem, but it seems that the library is no longer actively maintained.

## Possible Solutions

We can address this issue using one of the following approaches:

1. **Rewrite the Entire Library**

   - This would take a considerable amount of time.
   - It provides the potential for further development and customization.

2. **Apply a Patch**

   - The bug can be fixed by changing a single line of code in the library.
   - We already have a patches to this library

3. **Ignore the Bug**

   - The bug does not seem to impact the application's core functionality.

Based on these options, applying a patch appears to be the most efficient and practical solution at this time.

**Comment by blazejkustra** [comment]
I agree that adding a small patch for it should be the way to go, great investigation @sumo-slonik

**Comment by sumo-slonik** [comment]
@shubham1206agra In summary, the bug is related to a console error appearing when adding a new expense:

https://github.com/user-attachments/assets/c4cec5a1-f2dd-49c1-a93a-90757ff547ee

**Comment by sumo-slonik** [comment]
> @mountiny @blazejkustra
> 
> Regarding the bug we discussed on Slack: https://swmansion.slack.com/archives/C049HHMV9SM/p1733893876395249
> 
> I have identified the cause of the bug, which is located in the **react-native-draggable-flatlist** library. After investigating, I discovered an open [issue](https://github.com/computerjazz/react-native-draggable-flatlist/issues/564) related to this problem, but it seems that the library is no longer actively maintained.
> 
> ## Possible Solutions
> We can address this issue using one of the following approaches:
> 
> 1. **Rewrite the Entire Library**
>    
>    * This would take a considerable amount of time.
>    * It provides the potential for further development and customization.
> 2. **Apply a Patch**
>    
>    * The bug can be fixed by changing a single line of code in the library.
>    * We already have a patches to this library
> 3. **Ignore the Bug**
>    
>    * The bug does not seem to impact the application's core functionality.
> 
> Based on these options, applying a patch appears to be the most efficient and practical solution at this time.

Additionally, to prove that version two would be the best choice, I would like to add that it is our patch that causes this consol warrning so I think it is better to fix it

**Comment by blazejkustra** [comment]
> Here is an issue for that #53759
> 
> @blazejkustra @sumo-slonik is this everything already migrated to reanimated? it feels like there should be more

There are couple things left:
- Migrate `SearchTypeMenuNarrow` and `BaseOverlay` to Reanimated (should be rather easy)
- Adjust `TS_CHEATSHEET.md` doc - Animated mention is no longer needed
- Add Animated to `restrictedImportPaths` in our eslint config
- Removal of unused `fade` function from `src/styles/utils/index.ts` file

Also, @sumo-slonik didn't migrate `TabSelector` to Reanimated as this component is deeply connected with `react-navigation`, which relies on `Animated` under the hood. We believe migrating it would require a significant refactor of the `TopTab.Navigator`. We’ll continue investigating and provide an update later this week.

All this can be done in a third and final PR, sounds good? @mountiny

**Comment by blazejkustra** [comment]
>> we also found the use of react-native-animatable would we like to swap that too with this PR?
>
> Lets leave that out for different issue, feel free to write a proposal in open source and tag design team

I was investigating `react-native-animatable` today, and come to a conclusion that it is just a wrapper of `Animated` with an easier declarative API. This means that we'll get the same results when migrating to Reanimated. 

Also, turns out animatable is not even in our `package.json` 😅 It's just a dependency of `react-native-modal` (that we are currently migrating in another issue), so we shouldn't use it in the project in the first place.

Animatable is used in just two components (`BackgroundImage` and `AnimatedStep`), and it would be great to migrate it to Reanimated because of two reasons:
- Remove unnecessary dependency in the future (with two patches) 
- For overall consistency as reanimated is used in all other components

@mountiny do you think we need another proposal for it? Or we could do it as part of this issue? 🙏

**Comment by mountiny** [comment]
Lets use patch

Lets do this:
> Migrate SearchTypeMenuNarrow and BaseOverlay to Reanimated (should be rather easy)

in one PR

And the rest in another PR

> do you think we need another proposal for it? Or we could do it as part of this issue (or create a follow up issue)?

No I think we can cover it here, its good to use just one library for animations

Please investigate what it would take for the topBar nav to migrate too, but if its too complex or requires patches maybe we can skip

**Comment by stephanieelliott** [comment]
Ok so it sounds like we are waiting for 2 more PRs on this before issuing the payment -- I am going to remove the payment labels, we can reapply and calculate the payment date again once those are merged.

**Comment by stephanieelliott** [comment]
Seems like we are waiting on 1 more PR here? Removing payment info for the 2nd PR now, this issue should auto-update with payment details once the third PR is merged.

**Comment by sumo-slonik** [comment]
@stephanieelliott 
I'm currently working on migrating the switch animations, and after that, there will be just one PR left.

**Comment by blazejkustra** [comment]
There is also one that migrates animatable to reanimated that I'm working on [here](https://github.com/Expensify/App/pull/54404), so there are three in total 😄

**Comment by sumo-slonik** [comment]
The draft PR with the last two components is ready. Once the changes with the switch animations are accepted, I will add another PR with the completed changes for the accounting tab and the animations for the switch there. We're getting closer to the end of this issue.

**Comment by sumo-slonik** [comment]
I have opened the PR, and it's ready for review.

**Comment by blazejkustra** [comment]
Update: I finally had time to finish animatable migration (I had ton of work related to vip-travel project).

I decided to split it into two PRs, as initial PR got too big:
- https://github.com/Expensify/App/pull/55415
- https://github.com/Expensify/App/pull/54404

**Comment by blazejkustra** [comment]
@shubham1206agra https://github.com/Expensify/App/pull/54404 is ready for your review 😄

**Comment by mountiny** [comment]
@blazejkustra I think there is one last PR that you work on here, right?

**Comment by blazejkustra** [comment]
Yep, [this](https://github.com/Expensify/App/pull/54404) is the last migration.

**Comment by blazejkustra** [comment]
Note for future self, next time we should create multiple sub-issues 😅

**Comment by mountiny** [comment]
We could this is more of a tracking issue

**Comment by stephanieelliott** [comment]
Summarizing payment on this issue, it includes the following pull requests: 

- [Migrate animated to reanimated Tooltip and Switch #53298](https://github.com/Expensify/App/issues/53298)
- [Migrate/textinput growlnotification attachmentmodal animte to reanimate #53321](https://github.com/Expensify/App/issues/53321)
- [fix metro config and patch draggable flat list #54245](https://github.com/Expensify/App/issues/54245)
- [Migrate last components to reanimated #55202](https://github.com/Expensify/App/issues/55202)
- [Migrate SignInPage to react-native-reanimated, remove useNativeDriver #55415](https://github.com/Expensify/App/issues/55415)

---

 - Contributor: @blazejkustra EC, no payment due
 - Contributor: @sumo-slonik EC, no payment due
 - Contributor+: @shubham1206agra $1250 via Upwork -- offer extended, please accept when you get a chance! https://www.upwork.com/nx/wm/offer/106076797
 
 Upwork job is here: https://www.upwork.com/jobs/~021889212108205489060

**Comment by stephanieelliott** [comment]
OH, thanks! Ok removing payment labels again.

**Comment by blazejkustra** [comment]
Last PR was just merged, so we can slowly wrap this issue up!

**Comment by mountiny** [comment]
@blazejkustra no more components to migrate then?

**Comment by mountiny** [comment]
@shubham1206agra @stephanieelliott I believe this is now ready for payment. @shubham1206agra can you please help by summarizing the list of the PRs you have reviewed. Since there were some regressions along the way, I think we should decrease the total award so the $1250 posted by @stephanieelliott before actually looks appropriate to me.

**Comment by mountiny** [comment]
@shubham1206agra thanks for the summary, I think $1500 is appropriate, this has been a long project

**Comment by stephanieelliott** [comment]
Summarizing payment on this issue: 

 - Contributor: N/A (expert contrib)
 - Contributor+: @shubham1206agra $1500 -- please request via ND! 
 
 No Upwork job
